pkg/api/types.go

```diff
@@ -1664,6 +1667,104 @@ type SecretList struct {
+	// and volumes
+	SELinuxOptions
+}
+
+// SELinuxOptions are the labels to be applied to the container
+type SELinuxOptions struct {
+	// User --security-opt="label:user:USER"
+	User string
+
+	// Role --security-opt="label:role:ROLE"
+	Role string
+
+	// Type --security-opt="label:type:TYPE"
+	Type string
+
+	// Level --security-opt="label:level:LEVEL"
+	Level string
+
+	// Disabled --security-opt="label:disable"
+	Disabled bool
+}
+
+// SecurityConstraints provides the constraints that at security context provider will
+// ensure that applied SecurityContext requests follow.  When a setting is provided in the SecurityConstraints
+// that conflicts with an actual request it will be implementation specific whether that request is
+// ignored and the container is still run (without the requested constraint) or if the container will be failed
+type SecurityConstraints struct {
+	// EnforcementPolicy will drive behavior for how the constraints are enforce
+	EnforcementPolicy SecurityConstraintPolicy
+
+	// AllowPrivileged indicates whether this context allows privileged mode containers
+	AllowPrivileged bool
+
+	// SELinux provides the security constraint options for selinux
+	SELinux SELinuxSecurityConstraints
+
+	// AllowCapabilities dictates if a container can request to add or drop capabilites
+	AllowCapabilities bool
+
+	// Capabilities represents, if AllowCapabilities is true, the caps that requests
+	// are allowed to add or drop.
+	Capabilities Capabilities
+
+	// DefaultSecurityContext is applied to any container that does not have a security context set.  It must
+	// also conform to the constraints defined in SecurityConstraints object
+	DefaultSecurityContext SecurityContext
+}
+
+// SELinuxSecurityConstraints defines what is allowed in SecurityContext requests with regards to SELinux label options
+// that are currently supported by docker
+type SELinuxSecurityConstraints struct {
+	// AllowUserLabel --security-opt="label:user:USER"
+	AllowUserLabel  bool
+
+	// AllowRoleLabel --security-opt="label:role:ROLE"
+	AllowRoleLabel bool
+
+	// AllowTypeLabel --security-opt="label:type:TYPE"
+	AllowTypeLabel  bool
+
+	// AllowLevelLabel --security-opt="label:level:LEVEL"
+	AllowLevelLabel bool
+
+	// AllowDisable --security-opt="label:disable"
+	AllowDisable    bool
+}
+
+// SecurityConstraintPolicy dictates how the security context provider should behave with regards to contexts that
+// do not meet the requirements of the policy.
+type SecurityConstraintPolicy string
+
+const (
+	// SecurityConstraintPolicyDisabled means that any containers that do not meet policy constraints are still
+	// allowed to run
+	SecurityConstraintPolicyDisabled = "Disabled"
+
+	// SecurityConstraintPolicyPermissive means that any containers that do not meet policy constraints will be
+	// modified to meet the constraints (settings will be changed) before they are run
+	SecurityConstraintPolicyPermissive = "Permissive"
```

Comment:
I think we're in a situation where they all read like this if a policy is broken and we operate in a "give it a try anyway" permissive mode:

1.  api server: logs warning about broken policy, still accepts update (possibly manipulates applied context here based on the default context provided in SecurityConstraints)
2.  kubelet: removes offending policy or sets to default (if api server doesn't).  Container may function correctly but probably not.

To keep it boxed to what we can currently do today with docker we're talking about something like this:

1.  Add/remove caps
    * cap request would be removed
    * any missing required caps (based on policy) would be added - note this is not shown in the current proposed types.  I would need to add a required cap list
2.  Set SELinux labels (user, role, type, level, disabled)
    * label request would be removed
    * default label policy would be applied if set
3.  Set AppArmor profile
    * profile request would be removed
    * profile policy would be applied if set
4.  Run as privileged container
    * priv set to false

Requests to drop capabilities seem pretty benign, no issues there.  Specific requests for privs beyond the default docker capabilities, labels, or app armor profile are likely to cause issues if rejected or replaced.